### PR TITLE
Scripts/Midsummer: Burning hot pole dance

### LINKED ISSUE #6876: Scripts/Spells: Burning Hot Pole Dance  achievement
http://http://old.wowhead.com/achievement=271

achievement don't work ..after i put the set on me i clicked on Ribbon Pole (http://old.wowhead.com/object=181605) and not working.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/64453-achievement-burning-hot-pole-dance?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Aokromes**
**Changes proposed**:
- Fix Achievement Burning hot pole dance

**Target branch(es)**: 335

**Issues addressed**: Closes  #6876

**Tests performed**: (Does it build, tested in-game, etc)
Compiles

By  zwerg, closes #6876

### CODE REVIEW

**Comment by Demonid** [comment]
While this does the job for getting the achievement, it's missing a lot of effects:

https://www.youtube.com/watch?v=PkmTt93Ucsg

```diff

From 0f74147d5af3382854f160c2a29d876a3e0faaa4 Mon Sep 17 00:00:00 2001
From: Demonid <Demonidpvp@gmail.com>
Date: Wed, 22 Jun 2016 20:38:26 -0700
Subject: [PATCH] [SpecWoW] Event/Midsummer: Achievement: Burning Hot Pole
 Dance

---
 .../world/midsummer_burning_hot_pole_dance.sql     | 36 +++++++++
 src/server/scripts/Spells/spell_holiday.cpp        | 45 +++++++++++
 src/server/scripts/World/go_scripts.cpp            | 33 ++++++++
 src/server/scripts/World/npcs_special.cpp          | 90 ++++++++++++++++++++++
 4 files changed, 204 insertions(+)
 create mode 100644 sql/custom/world/midsummer_burning_hot_pole_dance.sql

diff --git a/sql/custom/world/midsummer_burning_hot_pole_dance.sql b/sql/custom/world/midsummer_burning_hot_pole_dance.sql
new file mode 100644
index 0000000..1345edd
--- /dev/null
+++ b/sql/custom/world/midsummer_burning_hot_pole_dance.sql
@@ -0,0 +1,36 @@
+-- Achievement - Burning Hot Pole Dance
+SET @GO := 181605;
+SET @SPELL_DANCE_PERIODIC := 45406;
+SET @SPELL_HAS_CLOTH := 58933;
+SET @SPELL_CHANNEL_1 := 29705;
+SET @SPELL_CHANNEL_2 := 29726;
+SET @SPELL_CHANNEL_3 := 29727;
+SET @RIBBON_TRIGGER := 17066;
+SET @RIBBON_BUNNY := 25303;
+
+DELETE FROM `spell_dbc` WHERE `Id`=58934;
+INSERT INTO `spell_dbc` (`Id`,`Effect1`,`EffectImplicitTargetA1`,`EffectImplicitTargetB1`,`Comment`) VALUES
+(58934,3,1,0,'Burning Hot Pole Dance (Gives achievement)');
+
+UPDATE `gameobject_template` SET `AIName`='', `ScriptName`='go_ribbon_pole' WHERE `entry`=@GO;
+UPDATE `creature_template` SET `AIName`='', `InhabitType`=4, `ScriptName`='npc_ribbon_pole_debug_target' WHERE `entry`=@RIBBON_TRIGGER;
+UPDATE `creature_template` SET `AIName`='SmartAI' WHERE `entry`=@RIBBON_BUNNY;
+
+DELETE FROM `smart_scripts` WHERE `entryorguid`=@RIBBON_BUNNY AND `source_type`=0;
+INSERT INTO `smart_scripts` (`entryorguid`,`source_type`,`id`,`link`,`event_type`,`event_phase_mask`,`event_chance`,`event_flags`,`event_param1`,`event_param2`,`event_param3`,`event_param4`,`action_type`,`action_param1`,`action_param2`,`action_param3`,`action_param4`,`action_param5`,`action_param6`,`target_type`,`target_param1`, `target_param2`,`target_param3`,`target_x`,`target_y`,`target_z`,`target_o`,`comment`) VALUES
+(@RIBBON_BUNNY,0,0,0,60,0,100,0,5000,5000,10000,10000,11,46836,0,0,0,0,0,1,0,0,0,0,0,0,0,'Ribbon Pole Bunny - Update - Cast Ribbon Pole Firework Launcher Aura, Long'),
+(@RIBBON_BUNNY,0,1,0,60,0,100,0,0,0,10000,10000,11,46842,0,0,0,0,0,1,0,0,0,0,0,0,0,'Ribbon Pole Bunny - Update - Cast Ribbon Pole Firework Launcher Aura');
+
+DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId`=13 AND `SourceEntry` IN (@SPELL_CHANNEL_1,@SPELL_CHANNEL_2,@SPELL_CHANNEL_3);
+INSERT INTO `conditions` (`SourceTypeOrReferenceId`,`SourceGroup`,`SourceEntry`,`SourceId`,`ElseGroup`,`ConditionTypeOrReference`,`ConditionTarget`,`ConditionValue1`,`ConditionValue2`,`ConditionValue3`,`NegativeCondition`,`ErrorTextId`,`ScriptName`,`Comment`) VALUES
+(13,1,@SPELL_CHANNEL_1,0,0,31,0,3,@RIBBON_TRIGGER,0,0,0,'','Test Ribbon Pole Channel - Implicit target - Ribbon Pole Debug Target'),
+(13,1,@SPELL_CHANNEL_2,0,0,31,0,3,@RIBBON_TRIGGER,0,0,0,'','Test Ribbon Pole Channel - Implicit target - Ribbon Pole Debug Target'),
+(13,1,@SPELL_CHANNEL_3,0,0,31,0,3,@RIBBON_TRIGGER,0,0,0,'','Test Ribbon Pole Channel - Implicit target - Ribbon Pole Debug Target');
+
+DELETE FROM `spell_linked_spell` WHERE `spell_trigger`=29715;
+INSERT INTO `spell_linked_spell` (`spell_trigger`, `spell_effect`, `type`, `comment`) VALUES
+(29715, 29175, 0, 'Pole Channel trigger Experience buff on each proc');
+
+DELETE FROM `spell_script_names` WHERE `spell_id`=@SPELL_DANCE_PERIODIC;
+INSERT INTO `spell_script_names` (`spell_id`,`ScriptName`) VALUES
+(@SPELL_DANCE_PERIODIC,'spell_midsummer_ribbon_pole_visual');
diff --git a/src/server/scripts/Spells/spell_holiday.cpp b/src/server/scripts/Spells/spell_holiday.cpp
index 7fe1f54..261d8a2 100644
--- a/src/server/scripts/Spells/spell_holiday.cpp
+++ b/src/server/scripts/Spells/spell_holiday.cpp
@@ -1032,6 +1032,50 @@ class spell_midsummer_braziers_hit : public SpellScriptLoader
         }
 };
 
+// 45406 - Holiday - Midsummer, Ribbon Pole Periodic Visual
+enum SpellsRibbon
+{
+    GO_RIBBON_POLE        = 181605,
+    SPELL_DANCE_PERIODIC  = 45406,
+    SPELL_CHANNEL_1       = 29705,
+    SPELL_CHANNEL_2       = 29726,
+    SPELL_CHANNEL_3       = 29727
+};
+
+class spell_midsummer_ribbon_pole_visual : public SpellScriptLoader
+{
+    public:
+        spell_midsummer_ribbon_pole_visual() : SpellScriptLoader("spell_midsummer_ribbon_pole_visual") { }
+
+        class spell_midsummer_ribbon_pole_visual_AuraScript : public AuraScript
+        {
+            PrepareAuraScript(spell_midsummer_ribbon_pole_visual_AuraScript);
+
+            void OnPeriodic(AuraEffect const* /*aurEff*/)
+            {
+                Unit* target = GetTarget();
+                if (!target->FindNearestGameObject(GO_RIBBON_POLE,10.0f))
+                {
+                    target->RemoveAura(SPELL_CHANNEL_1);
+                    target->RemoveAura(SPELL_CHANNEL_2);
+                    target->RemoveAura(SPELL_CHANNEL_3);
+                    target->RemoveAura(SPELL_DANCE_PERIODIC);
+                }
+
+            }
+
+            void Register()
+            {
+                OnEffectPeriodic += AuraEffectPeriodicFn(spell_midsummer_ribbon_pole_visual_AuraScript::OnPeriodic, EFFECT_0, SPELL_AURA_PERIODIC_DUMMY);
+            }
+        };
+
+        AuraScript* GetAuraScript() const
+        {
+            return new spell_midsummer_ribbon_pole_visual_AuraScript();
+        }
+};
+
 void AddSC_holiday_spell_scripts()
 {
     // Love is in the Air
@@ -1062,4 +1106,5 @@ void AddSC_holiday_spell_scripts()
     new spell_brewfest_barker_bunny();
     // Midsummer Fire Festival
     new spell_midsummer_braziers_hit();
+    new spell_midsummer_ribbon_pole_visual();
 }
diff --git a/src/server/scripts/World/go_scripts.cpp b/src/server/scripts/World/go_scripts.cpp
index ef4a2b0..d0ec9ca 100644
--- a/src/server/scripts/World/go_scripts.cpp
+++ b/src/server/scripts/World/go_scripts.cpp
@@ -1196,6 +1196,38 @@ public:
     }
 };
 
+/*######
+## go_ribbon_pole
+######*/
+
+enum RibbonPole
+{
+    NPC_RIBBON_TRIGGER  = 17066,
+    NPC_RIBBON_BUNNY    = 25303,
+    SPELL_CHANNEL_1     = 29705,
+    SPELL_CHANNEL_2     = 29726,
+    SPELL_CHANNEL_3     = 29727
+};
+
+class go_ribbon_pole : public GameObjectScript
+{
+public:
+    go_ribbon_pole() : GameObjectScript("go_ribbon_pole") { }
+
+    bool OnGossipHello(Player* player, GameObject* go)
+    {
+        if (!player->HasAura(SPELL_CHANNEL_1) && !player->HasAura(SPELL_CHANNEL_2) && !player->HasAura(SPELL_CHANNEL_3))
+        {
+            player->SummonCreature(NPC_RIBBON_TRIGGER, go->GetPositionX(), go->GetPositionY(), go->GetPositionZ() + 6.0f, TEMPSUMMON_MANUAL_DESPAWN);
+
+            if (!player->FindNearestCreature(NPC_RIBBON_BUNNY,20.0f))
+                player->SummonCreature(NPC_RIBBON_BUNNY, go->GetPositionX(), go->GetPositionY(), go->GetPositionZ(), TEMPSUMMON_MANUAL_DESPAWN);
+        }
+
+        return true;
+    }
+};
+
 void AddSC_go_scripts()
 {
     new go_cat_figurine();
@@ -1231,4 +1263,5 @@ void AddSC_go_scripts()
     new go_veil_skith_cage();
     new go_frostblade_shrine();
     new go_midsummer_bonfire();
+    new go_ribbon_pole;
 }
diff --git a/src/server/scripts/World/npcs_special.cpp b/src/server/scripts/World/npcs_special.cpp
index 07dab1d..72f740b 100644
--- a/src/server/scripts/World/npcs_special.cpp
+++ b/src/server/scripts/World/npcs_special.cpp
@@ -2448,6 +2448,95 @@ class npc_stable_master : public CreatureScript
         }
 };
 
+/*######
+## npc_ribbon_pole_debug_target
+######*/
+
+enum eRibbonPole
+{
+    NPC_RIBBON_POLE_FIRE_SPIRAL  = 25303,
+    SPELL_CHANNEL_1              = 29705,
+    SPELL_CHANNEL_2              = 29726,
+    SPELL_CHANNEL_3              = 29727,
+    SPELL_RIBBON_POLE_DANCER     = 45406,
+    SPELL_HAS_CLOTH              = 58933,
+    SPELL_GIVE_ACHIEVEMENT       = 58934
+};
+
+class npc_ribbon_pole_debug_target : public CreatureScript
+{
+public:
+    npc_ribbon_pole_debug_target() : CreatureScript("npc_ribbon_pole_debug_target") { }
+
+    CreatureAI* GetAI(Creature* creature) const override
+    {
+        return new npc_ribbon_pole_debug_targetAI(creature);
+    }
+
+    struct npc_ribbon_pole_debug_targetAI : ScriptedAI
+    {
+        npc_ribbon_pole_debug_targetAI(Creature* creature) : ScriptedAI(creature) {}
+
+        uint32 Timer;
+        ObjectGuid PlayerGUID;
+        bool SpinStarted;
+
+        void IsSummonedBy(Unit* owner)
+        {
+            switch(urand(1,3))
+            {
+                case 1:
+                    owner->CastSpell(me,SPELL_CHANNEL_1);
+                    break;
+                case 2:
+                    owner->CastSpell(me,SPELL_CHANNEL_2);
+                    break;
+                case 3:
+                    owner->CastSpell(me,SPELL_CHANNEL_3);
+                    break;
+                default:
+                    break;
+            }
+
+            SpinStarted  = false;
+            Timer = 5000;
+            PlayerGUID = owner->GetGUID();
+        }
+
+        void UpdateAI(uint32 diff) override
+        {
+            if (Timer <= diff)
+            {
+                if (!SpinStarted)
+                {
+                    if (Player* owner = ObjectAccessor::GetPlayer(*me, PlayerGUID))
+                        owner->CastSpell(owner,SPELL_RIBBON_POLE_DANCER,true);
+                    SpinStarted = true;
+                    Timer = 57000;
+                }
+
+                else
+                {
+                    if (Player* owner = ObjectAccessor::GetPlayer(*me, PlayerGUID))
+                        if (owner->HasAura(SPELL_HAS_CLOTH))
+                            owner->CastSpell(owner,SPELL_GIVE_ACHIEVEMENT,true);
+                }
+
+            }
+            else Timer -= diff;
+
+            if (SpinStarted && !me->HasAura(SPELL_CHANNEL_1) && !me->HasAura(SPELL_CHANNEL_2) && !me->HasAura(SPELL_CHANNEL_3))
+            {
+                if(Creature* bunny = me->FindNearestCreature(NPC_RIBBON_POLE_FIRE_SPIRAL,30.0f))
+                    bunny->DespawnOrUnsummon();
+
+                me->DespawnOrUnsummon();
+            }
+        }
+    };
+
+};
+
 void AddSC_npcs_special()
 {
     new npc_air_force_bots();
@@ -2472,4 +2561,5 @@ void AddSC_npcs_special()
     new npc_spring_rabbit();
     new npc_imp_in_a_ball();
     new npc_stable_master();
+    new npc_ribbon_pole_debug_target();
 }
-- 
1.8.3.msysgit.0

**Comment by joschiwald** [comment]
Better place is spell_holiday.cpp instead of spell_generic.cpp

File: src/server/scripts/Spells/spell_holiday.cpp [hunk 1]
```diff
@@ -1183,6 +1183,75 @@ class spell_midsummer_braziers_hit : public SpellScriptLoader
...
+class spell_gen_ribbon_pole_dancer_check : public SpellScriptLoader
+{
+    public:
+        spell_gen_ribbon_pole_dancer_check() : SpellScriptLoader("spell_gen_ribbon_pole_dancer_check") { }
+
+        class spell_gen_ribbon_pole_dancer_check_AuraScript : public AuraScript
+        {
+            PrepareAuraScript(spell_gen_ribbon_pole_dancer_check_AuraScript);
+
+            bool Validate(SpellEntry const* /*spell*/)
```
Comment (Keader) [inline src/server/scripts/Spells/spell_holiday.cpp]: missing a override here :)

File: src/server/scripts/Spells/spell_holiday.cpp [hunk 2]
```diff
@@ -1183,6 +1183,75 @@ class spell_midsummer_braziers_hit : public SpellScriptLoader
...
+
+                    // reward achievement criteria
+                    if (aur->GetMaxDuration() == 3600000 && target->HasAura(SPELL_HAS_FULL_MIDSUMMER_SET))
+                        target->CastSpell(target, SPELL_BURNING_HOT_POLE_DANCE, true);
+                }
+                else
+                    target->AddAura(SPELL_RIBBON_DANCE, target);
+            }
+
+            void Register()
```
Comment (Keader) [inline src/server/scripts/Spells/spell_holiday.cpp]: missing override

File: src/server/scripts/Spells/spell_holiday.cpp [hunk 3]
```diff
@@ -1183,6 +1183,75 @@ class spell_midsummer_braziers_hit : public SpellScriptLoader
...
+            {
+                OnEffectPeriodic += AuraEffectPeriodicFn(spell_gen_ribbon_pole_dancer_check_AuraScript::PeriodicTick, EFFECT_0, SPELL_AURA_PERIODIC_DUMMY);
+            }
+        };
+
+        AuraScript* GetAuraScript() const
```
Comment (Keader) [inline src/server/scripts/Spells/spell_holiday.cpp]: missing override

File: src/server/scripts/Spells/spell_holiday.cpp (hunk 1 above)
Comment (Demonid) [inline src/server/scripts/Spells/spell_holiday.cpp]: bool Validate(SpellEntry const* /*spell*/) should be changed for SpellInfo

Comment (Demonid) [inline src/server/scripts/Spells/spell_holiday.cpp]: SpellEntry should be changed for SpellInfo

File: src/server/scripts/Spells/spell_holiday.cpp [hunk 4]
```diff
@@ -1183,6 +1183,75 @@ class spell_midsummer_braziers_hit : public SpellScriptLoader
...
+
+            void PeriodicTick(AuraEffect const* /*aurEff*/)
+            {
+                Unit* target = GetTarget();
+
+                if (!target)
+                    return;
+
+                // check if aura needs to be removed
+                if (!target->FindNearestGameObject(GO_RIBBON_POLE, 20.0f) || !target->HasUnitState(UNIT_STAT_CASTING))
```
Comment (Demonid) [inline src/server/scripts/Spells/spell_holiday.cpp]: UNIT_STAT_CASTING > UNIT_STATE_CASTING

**Comment by Keader** [comment]
Aokromes, ignore it

@Demonid

Why you not use ```taskscheduler``` instead of ```if (Timer <= diff)```
code become more readable :)

File: src/server/scripts/Spells/spell_holiday.cpp [hunk 5]
```diff
@@ -1183,6 +1183,73 @@ class spell_midsummer_braziers_hit : public SpellScriptLoader
...
+                    return false;
+                return true;
+            }
+
+            void PeriodicTick(AuraEffect const* /*aurEff*/)
+            {
+                Unit* target = GetTarget();
+
+                if (!target)
+                    return;
```
Comment (joschiwald) [inline src/server/scripts/Spells/spell_holiday.cpp]: target never null

File: src/server/scripts/Spells/spell_holiday.cpp [hunk 6]
```diff
@@ -1183,6 +1183,73 @@ class spell_midsummer_braziers_hit : public SpellScriptLoader
...
+                {
+                    target->InterruptNonMeleeSpells(false);
+                    target->RemoveAurasDueToSpell(GetId());
+                    return;
+                }
+
+                // set xp buff duration
+                if (Aura* aur = target->GetAura(SPELL_RIBBON_DANCE))
+                {
+                    aur->SetMaxDuration(aur->GetMaxDuration() >= 3600000 ? 3600000 : aur->GetMaxDuration() + 180000);
```
Comment (joschiwald) [inline src/server/scripts/Spells/spell_holiday.cpp]: `aur->SetMaxDuration(std::min(3600000, aur->GetMaxDuration() + 180000));`
looks better

**Pull request merged**